# Практическая работа 12: Исследование работы LLM в контексте работы с собственными числами матрицы

Ваше ФИО:

Тимошкин Роман Вячеславович

Что хотим сегодня изучить:
1. Реализации методов поиска собственных значений и собственных векторов на Python
2. Метод главных компонент, который много где используется сам по себе и основан на работе с собственными значениями матриц
3. Метод главных компонент для оптимизации (уменьшения размеров) модели LLM
4. Общение с AI-ассистентами (готовыми LLM) для исследования (изучения) новой предметной области (пункт 3 из списка ниже)
5. Научиться писать промпты к LLM так, чтобы получать адекватные и неадекватные результаты (галлюцинации)
6. Находить и устранять галлюцинации LLM в программировании и исследовании, когда предметная область Вам уже знакома (пункт 2 из списка ниже)
7. Научиться оптимизировать модели LLM и проверять их точность (пункт 1 из списка ниже)

В контексте использования large language model (LLM) термин "исследование" обычно относится к процессу изучения, разработки или применения LLM для достижения научных, технических или академических целей. Он может иметь несколько значений в зависимости от конкретного контекста:

1. **Разработка модели**: Исследование включает в себя создание или совершенствование LLM, включая проектирование архитектуры, оптимизацию алгоритмов обучения или повышение производительности (например, точности, результативности или обобщения). Этим часто занимаются исследователи ИИ или инженеры.

2. **Изучение возможностей**: Исследование может означать изучение того, что могут делать LLM, например, тестирование их способности решать конкретные задачи (например, понимать естественный язык, рассуждать или генерировать код) или оценку их ограничений и предубеждений.

3. **Применение к предметной области**: Использование LLMS для поддержки исследований в других областях, таких как анализ научной литературы, генерация гипотез или автоматизация обработки данных в таких дисциплинах, как медицина, физика или социальные науки.

4. **Влияние на этику и общество**: Изучение последствий внедрения LLM, включая справедливость, безопасность, риски дезинформации или экологические издержки, связанные с обучением и выводами.

Например, исследователь может использовать LLM для анализа обширных наборов данных для исследования (прикладной работы) или для экспериментов с тонкой настройкой, чтобы улучшить производительность при выполнении узкоспециализированной задачи (разработка). Термин является широким, но обычно подразумевает систематическое изучение или экспериментирование с участием LLM.

### **План работы (разделы ноутбука)**
1. Выбрать один или несколько AI-ассистентов из списка ниже и перечислить их ниже.
2. Написать промпт (желательно один, чтобы можно было корректнее сравнить результат) на любой языке, на котором Вы ожидаете получить лучший результат, в результате которого выбрать модель с весами LLM для скачивания. Далее будем работать с этой моделью как с матрицей. Сравните ответы полученные от разных LLM в таблице.
3. После этого с помощью того же или нового набора AI-ассистентов, получите сниппет для скачивания и начала работы с моделью. Вставьте сниппеты в соответствующие ячейки. Запустите ячейки и сравните полученные результаты, не внося дополнительных изменений в код, если он не работает.
4. Если код после запуска не работает, получите у соответствующего AI-ассистента рекомендации по исправлению кода. Вставьте исправленный код в новую соответствующую ячейку. Запустите код и в случае, если код снова не работает, получите новые исправления. Продолжайте либо, пока код не заработает, либо пока AI-ассистент не начнёт слишком сильно галлюцинировать. Протоколируйте все диалоги в соответствующие ячейки. Проведите анализ полученных результатов в текущем разделе.
5. Напишите собственные реализации методов нахождения собственных чисел и собственных векторов матриц, используя вызовы готовых реализаций в стандартных библиотеках или подробные описания.
6. Заставьте AI-агентов написать сниппеты для тех же методов. Сравните на любой матрце.
7. Исследуйте метод главных компонент для оптимизации размера выбранной модели при помощи собственных значений и собственных векторов. Сравните результаты работы модели до и после оптимизации.
8. Заключение.

---

## Раздел 1. Выбрать один или несколько AI-ассистентов из списка ниже и перечислить их ниже.

Список:

- [ChatGPT (OpenAI)](https://chatgpt.com/)
- [Claude (Anthropic)](https://claude.online/chat)
- [Gemini (Google)](https://gemini.google.com/app)
- [Grok (xAI)](https://grok.com/)
- [Perplexity](https://www.perplexity.ai/)
- [DeepSeek](https://chat.deepseek.com/)
- [Qwen](https://chat.qwen.ai/)
- [Llama (Meta AI)](https://huggingface.co/chat/) c учетной записью от Hugging Face
- [Mistral](https://mistral.ai/)
- [OpenAssistant](https://open-assistant.io/chat/)
- [GigaChat (Sber)](https://giga.chat/#chat)

- [Microsoft Copilot](https://copilot.microsoft.com/chats/)
- [Tabnine](https://www.tabnine.com/pricing/) - платно, но есть триал
- [Amazon Q Developer](https://aws.amazon.com/ru/q/developer/)
- [GitHub Copilot](https://github.com/copilot)

Вы также можете использовать модели, которые не попали в список, но обязательно указывайте их источник, название и прочие аттрибуты.

---

## Раздел 2. Написать промпт, отправить AI-ассистентам, сравнить результаты
2.1. Написать промпт (желательно один, чтобы можно было корректнее сравнить результат) на любой языке, на котором Вы ожидаете получить лучший результат, в результате которого **выбрать** модель с весами LLM для скачивания (желательно выбирать небольшие модели, чтобы скачивать было не много и матрицы были не слишком большими). Далее будем работать с этой моделью как с матрицей.

2.2. Отправить промпт AI-ассистентам, выбранным в разделе 1.

2.3. Сравните ответы полученные от разных LLM в таблице.

2.4. Выбрать модель для скачивания

### 2.1 Использованный промпт:

Посоветуй, пожалуйста, несколько небольших по размеру (до 1 миллиарда параметров) предобученных моделей LLM (трансформеров), которые можно легко скачать с Hugging Face. Укажи их названия (идентификаторы на Hugging Face) и краткое описание или область применения. Мне нужна модель для экспериментов с матрицами весов.

### 2.3. Сравнение результатов

| **Ассистент** | **Ответ** |**Ваш комментарий**|
|---|---|---|
|Gemini (Google)| Предложил `TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T`, `EleutherAI/pythia-410m`, `distilbert/distilgpt2`. Дал краткие описания и идентификаторы HF. | Ответ релевантный, модели подходят по размеру. |
|ChatGPT (OpenAI)| Предложил `tiiuae/falcon-rw-1b`, `EleutherAI/gpt-neo-1.3B`, `facebook/opt-1.3b`. Предоставил описания и идентификаторы HF. | Ответ также хороший, предложил разнообразные варианты. |
|DeepSeek| Предложил `openai-community/gpt2`, `distilbert/distilgpt2`, `huawei-noah/TinyBERT_General_4L_312D`. Дал описания и ссылки на HF. | Дал самые легкие модели, указал идентификаторы для скачивания. |

### 2.4. Выбранная модель

Для дальнейшей работы выбрана модель `distilbert/distilgpt2`, так как она является стандартной небольшой моделью, предложенной всеми ассистентами, и хорошо подходит для экспериментов из-за своего относительно небольшого размера и известной архитектуры.

---

## Раздел 3. Скачивание модели с помощью сниппетов от AI-ассистентов
С помощью того же или нового набора AI-ассистентов, получите сниппет для скачивания и начала работы с моделью. Вставьте сниппеты в соответствующие ячейки. Запустите ячейки и сравните полученные результаты, не внося дополнительных изменений в код, если он не работает.

3.1. Список используемых AI

3.2. Сниппеты для скачивания выбранной модели

3.3. Запуск сниппетов

3.4. Анализ результатов

### 3.1. Список используемых AI

Использовались те же ассистенты: Gemini, ChatGPT, GigaChat.

Промпт:

Напиши Python код с использованием библиотеки transformers от Hugging Face для загрузки модели 'distilbert/distilgpt2'. Мне нужно получить доступ к матрицам весов одного из её слоев (например, первого слоя трансформера) как к numpy-массивам.

### 3.2. Сниппеты для скачивания выбранной модели

Модель 1: Gemini (Google)

In [1]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
import numpy as np

# Идентификатор модели на Hugging Face
model_name = "distilbert/distilgpt2"

try:
    # Загрузка модели
    # use_safetensors=True рекомендуется для более безопасной и быстрой загрузки
    model = AutoModelForCausalLM.from_pretrained(model_name, use_safetensors=True)

    print(f"Модель '{model_name}' успешно загружена.")
    print(f"Общее количество параметров модели: {sum(p.numel() for p in model.parameters())}")

    # Структура моделей на основе GPT-2 (включая DistilGPT2) обычно
    # содержит слои трансформера в атрибуте 'transformer.h'.
    # Получаем первый слой (индекс 0)
    if hasattr(model, 'transformer') and hasattr(model.transformer, 'h'):
        first_transformer_layer = model.transformer.h[0]
        print("\nДоступ к первому слою трансформера (индекс 0).")

        # Внутри слоя трансформера есть блоки внимания (attn) и MLP (mlp).
        # У них есть линейные слои с весами.
        # Получим, например, веса проекции запросов/ключей/значений (QKV)
        # и веса первого линейного слоя в MLP.

        # Веса QKV проекции в блоке внимания
        if hasattr(first_transformer_layer, 'attn') and hasattr(first_transformer_layer.attn, 'c_attn'):
            qkv_weights_tensor = first_transformer_layer.attn.c_attn.weight.data
            qkv_weights_numpy = qkv_weights_tensor.cpu().numpy() # Перемещаем на CPU и конвертируем в numpy
            print(f"\nВеса проекции QKV первого слоя attention:")
            print(f"  Тип: {type(qkv_weights_numpy)}")
            print(f"  Размерность (shape): {qkv_weights_numpy.shape}")
            # print(f"  Несколько значений: {qkv_weights_numpy.flatten()[:10]}") # Пример первых 10 значений

            qkv_bias_tensor = first_transformer_layer.attn.c_attn.bias.data
            qkv_bias_numpy = qkv_bias_tensor.cpu().numpy()
            print(f"Веса смещения (bias) проекции QKV первого слоя attention:")
            print(f"  Тип: {type(qkv_bias_numpy)}")
            print(f"  Размерность (shape): {qkv_bias_numpy.shape}")

        # Веса первого линейного слоя в блоке MLP
        if hasattr(first_transformer_layer, 'mlp') and hasattr(first_transformer_layer.mlp, 'c_fc'):
            ffn1_weights_tensor = first_transformer_layer.mlp.c_fc.weight.data
            ffn1_weights_numpy = ffn1_weights_tensor.cpu().numpy() # Перемещаем на CPU и конвертируем в numpy
            print(f"\nВеса первого линейного слоя (c_fc) MLP первого слоя:")
            print(f"  Тип: {type(ffn1_weights_numpy)}")
            print(f"  Размерность (shape): {ffn1_weights_numpy.shape}")
            # print(f"  Несколько значений: {ffn1_weights_numpy.flatten()[:10]}") # Пример первых 10 значений

            ffn1_bias_tensor = first_transformer_layer.mlp.c_fc.bias.data
            ffn1_bias_numpy = ffn1_bias_tensor.cpu().numpy()
            print(f"Веса смещения (bias) первого линейного слоя (c_fc) MLP первого слоя:")
            print(f"  Тип: {type(ffn1_bias_numpy)}")
            print(f"  Размерность (shape): {ffn1_bias_numpy.shape}")

        # Можно аналогично получить веса других частей слоя, например:
        # Веса выходной проекции attention: first_transformer_layer.attn.c_proj.weight.data.cpu().numpy()
        # Веса второго линейного слоя MLP: first_transformer_layer.mlp.c_proj.weight.data.cpu().numpy()
        # Веса нормализации слоя: first_transformer_layer.ln_1.weight.data.cpu().numpy(), etc.

    else:
        print("\nНе удалось получить доступ к слоям трансформера в ожидаемой структуре.")
        print("Возможно, структура модели отличается. Изучите объект 'model'.")


except Exception as e:
    print(f"Произошла ошибка при загрузке или доступе к модели: {e}")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

Модель 'distilbert/distilgpt2' успешно загружена.
Общее количество параметров модели: 81912576

Доступ к первому слою трансформера (индекс 0).

Веса проекции QKV первого слоя attention:
  Тип: <class 'numpy.ndarray'>
  Размерность (shape): (768, 2304)
Веса смещения (bias) проекции QKV первого слоя attention:
  Тип: <class 'numpy.ndarray'>
  Размерность (shape): (2304,)

Веса первого линейного слоя (c_fc) MLP первого слоя:
  Тип: <class 'numpy.ndarray'>
  Размерность (shape): (768, 3072)
Веса смещения (bias) первого линейного слоя (c_fc) MLP первого слоя:
  Тип: <class 'numpy.ndarray'>
  Размерность (shape): (3072,)


Модель 2: ChatGPT (OpenAI)

In [2]:
import torch
import numpy as np
from transformers import AutoModelForCausalLM, AutoTokenizer

# Загрузка модели и токенизатора
model_name = 'distilbert/distilgpt2'
model = AutoModelForCausalLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Убедимся, что модель в режиме оценки
model.eval()

# Извлечение первого слоя трансформера
# Для GPT-подобной модели это модельный блок в transformer.h
first_block = model.transformer.h[0]

# Получим веса attention-слоя (например, матрицы для Q, K, V)
# Они находятся в: c_attn.weight
qkv_weights = first_block.attn.c_attn.weight.detach().cpu().numpy()

# Альтернатива — вес output-проекции attention-а
attn_out_proj_weights = first_block.attn.c_proj.weight.detach().cpu().numpy()

# Также можно получить веса feed-forward слоя
ffn_weights = first_block.mlp.c_fc.weight.detach().cpu().numpy()

# Пример вывода форм
print("QKV weight shape:", qkv_weights.shape)
print("Attention output proj weight shape:", attn_out_proj_weights.shape)
print("Feedforward weight shape:", ffn_weights.shape)

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

QKV weight shape: (768, 2304)
Attention output proj weight shape: (768, 768)
Feedforward weight shape: (768, 3072)


Модель 3: DeepSeek

In [11]:
from transformers import AutoModel
import numpy as np

# Загружаем модель и её конфигурацию
model_name = "distilgpt2"
model = AutoModel.from_pretrained(model_name)

# Переводим модель в режим оценки (отключаем dropout и т.д.)
model.eval()

# Получаем первый слой трансформера
# У DistilGPT2 слои находятся в `transformer.h`
first_layer = model.transformer.h[0]

# Доступ к матрицам весов первого слоя:
# 1. Матрицы внимания (query, key, value, output)
q_weights = first_layer.attention.q_lin.weight.data.numpy()  # (768, 768)
k_weights = first_layer.attention.k_lin.weight.data.numpy()  # (768, 768)
v_weights = first_layer.attention.v_lin.weight.data.numpy()  # (768, 768)
out_weights = first_layer.attention.out_lin.weight.data.numpy()  # (768, 768)

# 2. Полносвязные слои (двухслойный MLP)
fc1_weights = first_layer.ffn.lin1.weight.data.numpy()  # (3072, 768)
fc2_weights = first_layer.ffn.lin2.weight.data.numpy()  # (768, 3072)

# 3. Нормализация (LayerNorm)
ln1_weights = first_layer.sa_layer_norm.weight.data.numpy()  # (768,)
ln2_weights = first_layer.output_layer_norm.weight.data.numpy()  # (768,)

# Пример вывода формы матриц
print("Query weights shape:", q_weights.shape)
print("Key weights shape:", k_weights.shape)
print("Value weights shape:", v_weights.shape)
print("Output projection shape:", out_weights.shape)
print("FFN fc1 shape:", fc1_weights.shape)
print("FFN fc2 shape:", fc2_weights.shape)
print("LayerNorm 1 weights shape:", ln1_weights.shape)

AttributeError: 'GPT2Model' object has no attribute 'transformer'

### 3.4. Анализ результатов

Код, предоставленный Gemini и ChatGPT, успешно справился с задачей загрузки модели 'distilbert/distilgpt2' и извлечения матриц весов указанного слоя в виде numpy-массивов. Оба ассистента корректно использовали библиотеку `transformers` и методы для доступа к параметрам модели.

Код от DeepSeek изначально содержал ошибку (`AttributeError: 'GPT2Model' object has no attribute 'transformer'`). Это произошло из-за неверного пути доступа к слоям модели (`model.transformer.h[0]` вместо `model.h[0]` для `AutoModel`, или использования `AutoModelForCausalLM` как в других примерах, где путь `model.transformer.h[0]` корректен). Это типичный пример галлюцинации LLM, когда модель генерирует код, основанный на общих паттернах (многие трансформеры имеют атрибут `transformer`), но не учитывает специфику конкретного класса (`AutoModel` для `distilgpt2` в данном случае) или предложенного способа загрузки. После исправления пути доступа к слоям код заработал.

---

## Раздел 4. Первые галлюцинации
Если код после запуска не работает, получите у соответствующего AI-ассистента рекомендации по исправлению кода. Вставьте исправленный код в новую соответствующую ячейку. Запустите код и в случае, если код снова не работает, получите новые исправления. Продолжайте либо, пока код не заработает, либо пока AI-ассистент не начнёт слишком сильно галлюцинировать. Протоколируйте все диалоги в соответствующие ячейки. Проведите анализ полученных результатов в текущем разделе.

В случае, если все модели отработали безупречно, вернитесь к этому разделу после промта по написанию методов рассчёта собственных значений матрицы (раздел 6) или оптимизации выбранной модели LLM при помощи методов поиска собственных значений и собственных векторов матрицы весов. (раздел 7)

#### Справка: Когда LLM начинают галлюцинировать?

**LLMs склонны к галлюцинациям во время написания кода, главным образом, в таких ситуациях:**

1. **Неоднозначные или недостаточно определенные запросы**  
   - Если запрос расплывчатый ("напишите алгоритм быстрой сортировки"), то LLM может начать гадать и выдумывать детали.
2. **Редкие или незнакомые библиотеки, API или фреймворки**  
   - Если Вы спрашиваете о малоизвестном инструменте или совершенно новой версии библиотеки, LLM может "придумать" методы, классы или шаблоны использования на основе шаблонов, которые он видел в других местах.
3. **Междоменные запросы**  
   - При объединении технологий (например, "использовать модели SQLAlchemy внутри конвейера TensorFlow") могут возникнуть галлюцинации с API, которых на самом деле не существует.
4. **Устаревшие или будущие знания**  
   - Если что-то изменилось после завершения обучения модели (например, новые возможности версии Python), она может придумать поведение, которое "ожидает", основываясь на тенденциях.
5. **Сложные многоэтапные задачи**  
   - В более длинных цепочках (например, при полной настройке приложения) накапливаются ошибки - пропущенный импорт, неправильные типы, воображаемые функции и т.д.
6. **Когда вас просят оптимизировать, упростить или сильно обобщить**  
   - "Напишите наиболее эффективную систему кэширования" — это может привести к созданию идеализированных методов, которые звучат правдоподобно, но не являются реальными.
7. **Слишком самоуверенные формулировки в запросах**  
   - Если вы спросите "Дайте мне точный вызов API для..." с полной уверенностью, модель может выдать очень уверенный, но совершенно неверный ответ.

**Как распознать галлюцинации, когда LLMs пишут код:**

- **Сверьтесь с официальной документацией.**  
  Всегда сверяйте названия функций, сигнатуры методов и использование библиотек с реальной документацией.

- **Запустите и протестируйте код на ранней стадии.**  
  Даже небольшой пример (модульный тест, выполнение скрипта) может сразу выявить недостоверные детали.

- **Используйте средства проверки типов и линтеры.**  
  Такие инструменты, как "mypy", "pylint" или "eslint", часто обнаруживают (*выдуманные*) ошибочные API, неправильный импорт и т.д.

- **Задавайте дополнительные вопросы.**  
  Если вы не уверены, вы можете спросить у LLM: "Вы уверены, что `foo.bar()` существует? Можете ли вы показать документацию?". Но в последних версиях LLM поголовно научились нагло врать, наставать и выдумывать ссылки.

- **Будьте подозрительны к слишком точным ответам.**  
  Реальный код часто требует обработки ошибок, нестандартных ситуаций, шагов по настройке. Галлюцинированный код часто выглядит "слишком чистым".

- **Делайте подсказки конкретными.**  
  Сужение задачи снижает вероятность того, что модель ошибется в своих предположениях.

Модель 1: DeepSeek

In [12]:
from transformers import AutoModel
import numpy as np

# Загружаем модель
model_name = "distilgpt2"
model = AutoModel.from_pretrained(model_name)
model.eval()  # Переводим в режим оценки

# Получаем первый слой трансформера
# У GPT2 и DistilGPT2 слои находятся прямо в `.h`
first_layer = model.h[0]  # <-- Исправлено здесь!

# Доступ к матрицам весов:
# 1. Матрицы внимания (query, key, value, output)
q_weights = first_layer.attn.q_lin.weight.data.numpy()  # (768, 768)
k_weights = first_layer.attn.k_lin.weight.data.numpy()  # (768, 768)
v_weights = first_layer.attn.v_lin.weight.data.numpy()  # (768, 768)
out_weights = first_layer.attn.out_lin.weight.data.numpy()  # (768, 768)

# 2. Полносвязные слои (MLP)
fc1_weights = first_layer.mlp.lin1.weight.data.numpy()  # (3072, 768)
fc2_weights = first_layer.mlp.lin2.weight.data.numpy()  # (768, 3072)

# 3. Нормализация (LayerNorm)
ln1_weights = first_layer.ln_1.weight.data.numpy()  # (768,)
ln2_weights = first_layer.ln_2.weight.data.numpy()  # (768,)

# Вывод форм матриц
print("Query weights shape:", q_weights.shape)
print("Key weights shape:", k_weights.shape)
print("Value weights shape:", v_weights.shape)
print("Output projection shape:", out_weights.shape)
print("FFN fc1 shape:", fc1_weights.shape)
print("FFN fc2 shape:", fc2_weights.shape)
print("LayerNorm 1 weights shape:", ln1_weights.shape)

AttributeError: 'GPT2Attention' object has no attribute 'q_lin'

In [27]:
from transformers import AutoModel
import numpy as np
import torch

# Загружаем модель
model_name = "distilgpt2"
model = AutoModel.from_pretrained(model_name)
model.eval()

# Получаем первый слой трансформера
first_layer = model.h[0]

# 1. Веса механизма внимания
# У GPT2/DistilGPT2 Q, K, V объединены в одну матрицу c_attn
c_attn_weights = first_layer.attn.c_attn.weight.data  # (768, 2304 = 3*768)
c_attn_bias = first_layer.attn.c_attn.bias.data      # (2304,)

# Разделяем на Q, K, V
dim = c_attn_weights.shape[0]  # 768
q_weights = c_attn_weights[:, :dim].numpy()          # (768, 768)
k_weights = c_attn_weights[:, dim:2*dim].numpy()     # (768, 768)
v_weights = c_attn_weights[:, 2*dim:].numpy()        # (768, 768)

# Веса выходной проекции внимания
out_weights = first_layer.attn.c_proj.weight.data.numpy()  # (768, 768)

# 2. Веса MLP (полносвязные слои)
mlp_fc1 = first_layer.mlp.c_fc.weight.data.numpy()    # (3072, 768)
mlp_fc2 = first_layer.mlp.c_proj.weight.data.numpy()  # (768, 3072)

# 3. Веса LayerNorm
ln_1_weights = first_layer.ln_1.weight.data.numpy()   # (768,)
ln_2_weights = first_layer.ln_2.weight.data.numpy()   # (768,)

# Вывод форм матриц
print("Q weights shape:", q_weights.shape)
print("K weights shape:", k_weights.shape)
print("V weights shape:", v_weights.shape)
print("Output projection shape:", out_weights.shape)
print("MLP fc1 shape:", mlp_fc1.shape)
print("MLP fc2 shape:", mlp_fc2.shape)

target_matrix = q_weights

Q weights shape: (768, 768)
K weights shape: (768, 768)
V weights shape: (768, 768)
Output projection shape: (768, 768)
MLP fc1 shape: (768, 3072)
MLP fc2 shape: (3072, 768)


---

## Раздел 5. Набросаем код методов самостоятельно из готовой документации

Реализуйте расчёт собственных значений матриц весов любого слоя при помощи известных Вам методов самостоятельно, используя стандартные библиотеки Python.

Выполните это задание самостоятельно, а не при помощи ассистентов (**любых**, включая GitHub Copilot autocomplite, его можно отключить в настройках справа внизу, Google Colab Gemini тоже можнот отключить в настройках "Инструменты"-"ИИ-функции"-"Предлагать подсказки ИИ для дополнения кода"), чтобы можно было корректно сравнить результаты стандартных библиотек и сниппетов, которые предложат Вам ассистенты далее.


| Метод | Функция Python | Библиотека | Ссылка на документацию |
|---|---|---|---|
|Power Method (Dominant Eigenvalue) | Manual (template provided) | Custom/Numpy | [Tutorial](https://bvanderlei.github.io/jupyter-guide-to-linear-algebra/Approximating_Eigenvalues.html#power-method) or [template](https://en.wikipedia.org/wiki/Power_iteration#Numerical_example) — implement with `numpy.dot` |
|Shifted Power Method | Manual + `numpy.linalg.solve` | Numpy | [Tutorial](https://bvanderlei.github.io/jupyter-guide-to-linear-algebra/Approximating_Eigenvalues.html) |
|QR Algorithm (Full Spectrum) | `scipy.linalg.eig` or `numpy.linalg.eig` | Scipy/Numpy | [scipy.linalg.eig](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eig.html) / [numpy.linalg.eig](https://numpy.org/doc/stable/reference/generated/numpy.linalg.eig.html) |
|Full QR Decomposition (optional deeper understanding) | `numpy.linalg.qr` | Numpy | [numpy.linalg.qr](https://numpy.org/doc/stable/reference/generated/numpy.linalg.qr.html) |

In [21]:
import numpy as np

def power_iteration(matrix, num_simulations: int = 100):
    n = matrix.shape[0]
    # Инициализируем случайный вектор
    b_k = np.random.rand(n)

    for _ in range(num_simulations):
        # Вычисляем произведение матрицы на вектор
        b_k1 = np.dot(matrix, b_k)

        # Нормализуем вектор
        b_k1_norm = np.linalg.norm(b_k1)
        b_k = b_k1 / b_k1_norm

    # Собственное значение Рэлея
    eigenvalue = np.dot(b_k.T, np.dot(matrix, b_k)) / np.dot(b_k.T, b_k)
    return eigenvalue, b_k

# Генерация случайной матрицы весов (например, 4x4)
np.random.seed(42)
W = np.random.randn(4, 4)

dominant_eigenvalue, dominant_eigenvector = power_iteration(A)

# Расчет собственных значений и векторов
eigenvalues_np, eigenvectors_np = np.linalg.eig(W)

# Вывод результатов
print("Матрица весов W:\n", W)
print("\nСобственные значения:\n", eigenvalues_np)
print("\nСобственные векторы (столбцы):\n", eigenvectors_np)

Матрица весов W:
 [[ 0.49671415 -0.1382643   0.64768854  1.52302986]
 [-0.23415337 -0.23413696  1.57921282  0.76743473]
 [-0.46947439  0.54256004 -0.46341769 -0.46572975]
 [ 0.24196227 -1.91328024 -1.72491783 -0.56228753]]

Собственные значения:
 [ 1.43750501+0.j         -0.79407997+0.j         -0.70327653+1.06686096j
 -0.70327653-1.06686096j]

Собственные векторы (столбцы):
 [[ 0.70678858+0.j         -0.38383401+0.j         -0.37825022-0.44822856j
  -0.37825022+0.44822856j]
 [-0.18167634+0.j          0.48849804+0.j          0.02871948-0.49840163j
   0.02871948+0.49840163j]
 [-0.36764002+0.j         -0.51649525+0.j         -0.03351373+0.1010031j
  -0.03351373-0.1010031j ]
 [ 0.57644117+0.j          0.58929939+0.j          0.6288594 +0.j
   0.6288594 -0.j        ]]


---

## Раздел 6. AI-ассистенты пишут код для расчёта собственных значений и векторов матриц
6.1. Напишите промпт для получения сниппетов методов расчёта собственных значений и векторов матриц.

6.2. Сниппеты для расчёта собственных значений и векторов матриц

6.3. Запуск сниппетов против любой выбранной Вами матрицы

6.4. Анализ результатов

### 6.1 Ваш промпт

Напиши функции Python для вычисления:
1. Доминантного собственного значения и соответствующего собственного вектора матрицы с использованием степенного метода (power iteration).
2. Всех собственных значений и собственных векторов матрицы с использованием функции из библиотеки NumPy или SciPy.

Продемонстрируй их работу на небольшой тестовой матрице. Используй numpy для матричных операций.

### 6.2. Сниппеты для расчёта собственных значений и векторов матриц

Модель 1: Gemini (Google)

In [20]:
import numpy as np

def power_iteration_gemini(matrix, num_iterations=1000, tolerance=1e-6):
    """
    Вычисляет доминантное собственное значение и собственный вектор
    матрицы с использованием степенного метода.

    Аргументы:
        matrix (np.ndarray): Квадратная NumPy матрица.
        num_iterations (int): Максимальное количество итераций.
        tolerance (float): Критерий остановки по норме разности векторов.

    Возвращает:
        tuple: (доминантное_собственное_значение, доминантный_собственный_вектор)
               или (None, None), если метод не сошелся в пределах итераций.
    """
    n = matrix.shape[0]
    if matrix.shape[1] != n:
        raise ValueError("Матрица должна быть квадратной.")

    # 1. Инициализировать случайный начальный вектор
    # Убедимся, что вектор не нулевой
    eigenvector = np.random.rand(n)
    # Нормализовать вектор
    eigenvector = eigenvector / np.linalg.norm(eigenvector)

    dominant_eigenvalue = None

    # Итеративный процесс
    for _ in range(num_iterations):
        # 2. Вычислить произведение матрицы на вектор (Av_k)
        Av = np.dot(matrix, eigenvector)

        # 3. Оценить собственное значение (используя норму Av_k)
        # Норма Av_k сходится к абсолютному значению доминантного собственного значения
        Av_norm = np.linalg.norm(Av)

        # 4. Нормализовать новый вектор (v_{k+1} = Av_k / ||Av_k||)
        new_eigenvector = Av / Av_norm

        # 5. Проверить сходимость
        # Сравниваем новый вектор с предыдущим
        if np.linalg.norm(new_eigenvector - eigenvector) < tolerance:
             # Вычисляем собственное значение более точно с помощью отношения Рэлея
             # для нормализованного вектора
             dominant_eigenvalue = np.dot(new_eigenvector.T, np.dot(matrix, new_eigenvector))
             return dominant_eigenvalue, new_eigenvector

        # Обновить вектор для следующей итерации
        eigenvector = new_eigenvector

    # Если метод не сошелся за num_iterations
    print(f"Предупреждение: Степенной метод не сошелся за {num_iterations} итераций.")
    # В случае несходимости вернем последнюю оценку и векторы
    # Или можно вернуть None, None
    dominant_eigenvalue = np.dot(eigenvector.T, np.dot(matrix, eigenvector)) # Последняя оценка
    return dominant_eigenvalue, eigenvector # Возвращаем последний вектор и его оценку с/з

def get_all_eigen_np(matrix):
    """
    Вычисляет все собственные значения и собственные векторы матрицы
    с использованием функции из библиотеки NumPy.

    Аргументы:
        matrix (np.ndarray): Квадратная NumPy матрица.

    Возвращает:
        tuple: (собственные_значения, собственные_векторы)
               где собственные_значения - массив собственных значений,
               а собственные_векторы - матрица, столбцами которой являются
               соответствующие собственные векторы.
    """
    if matrix.shape[0] != matrix.shape[1]:
        raise ValueError("Матрица должна быть квадратной.")

    # Используем функцию np.linalg.eig()
    eigenvalues, eigenvectors = np.linalg.eig(matrix)

    return eigenvalues, eigenvectors

# --- Демонстрация работы ---

# Создадим небольшую тестовую матрицу
# Например, матрица 2x2 с известными собственными значениями
# Матрица A = [[4, 1], [2, 3]]
# Собственные значения: lambda1 = 5, lambda2 = 2
test_matrix = np.array([[4, 1],
                        [2, 3]])

print("Тестовая матрица:")
print(test_matrix)
print("-" * 30)

# Демонстрация степенного метода
print("Результат степенного метода (нахождение доминантного собственного значения/вектора):")
dominant_val_pi, dominant_vec_pi = power_iteration_gemini(test_matrix)

if dominant_val_pi is not None:
    print(f"  Доминантное собственное значение: {dominant_val_pi}")
    print(f"  Соответствующий собственный вектор: {dominant_vec_pi}")

    # Проверка: A * v = lambda * v
    Av = np.dot(test_matrix, dominant_vec_pi)
    lambdav = dominant_val_pi * dominant_vec_pi
    print("\n  Проверка (A * v vs lambda * v):")
    print(f"    A * v: {Av}")
    print(f"    lambda * v: {lambdav}")
    print(f"    Разница (норма): {np.linalg.norm(Av - lambdav)}") # Должна быть близка к нулю

print("-" * 30)

# Демонстрация вычисления всех собственных значений/векторов с помощью NumPy
print("Результат с использованием np.linalg.eig (все собственные значения/вектора):")
all_vals, all_vecs = get_all_eigen_np(test_matrix)

print(f"  Все собственные значения: {all_vals}")
print("  Соответствующие собственные векторы (по столбцам):")
print(all_vecs)

# Проверка для всех пар (опционально)
print("\n  Проверка для всех пар (A * v[:, i] vs lambda[i] * v[:, i]):")
for i in range(len(all_vals)):
    eigenvalue = all_vals[i]
    eigenvector = all_vecs[:, i] # Получаем i-й столбец как вектор

    Av = np.dot(test_matrix, eigenvector)
    lambdav = eigenvalue * eigenvector

    print(f"    Пара {i+1}:")
    print(f"      Собственное значение: {eigenvalue}")
    print(f"      Собственный вектор: {eigenvector}")
    print(f"      A * v: {Av}")
    print(f"      lambda * v: {lambdav}")
    print(f"      Разница (норма): {np.linalg.norm(Av - lambdav)}") # Должна быть близка к нулю
    print("-" * 10)

# Попробуем другую матрицу 3x3
test_matrix_3x3 = np.array([[3, -1, 0],
                           [-1, 2, -1],
                           [0, -1, 3]])

print("\n" + "=" * 30)
print("Тестовая матрица 3x3:")
print(test_matrix_3x3)
print("-" * 30)

print("Результат степенного метода (нахождение доминантного собственного значения/вектора):")
dominant_val_pi_3x3, dominant_vec_pi_3x3 = power_iteration_gemini(test_matrix_3x3)

if dominant_val_pi_3x3 is not None:
    print(f"  Доминантное собственное значение: {dominant_val_pi_3x3}")
    print(f"  Соответствующий собственный вектор: {dominant_vec_pi_3x3}")
    # Проверка
    Av_3x3 = np.dot(test_matrix_3x3, dominant_vec_pi_3x3)
    lambdav_3x3 = dominant_val_pi_3x3 * dominant_vec_pi_3x3
    print("\n  Проверка (A * v vs lambda * v):")
    print(f"    A * v: {Av_3x3}")
    print(f"    lambda * v: {lambdav_3x3}")
    print(f"    Разница (норма): {np.linalg.norm(Av_3x3 - lambdav_3x3)}")

print("-" * 30)
print("Результат с использованием np.linalg.eig (все собственные значения/вектора):")
all_vals_3x3, all_vecs_3x3 = get_all_eigen_np(test_matrix_3x3)

print(f"  Все собственные значения: {all_vals_3x3}")
print("  Соответствующие собственные векторы (по столбцам):")
print(all_vecs_3x3)

Тестовая матрица:
[[4 1]
 [2 3]]
------------------------------
Результат степенного метода (нахождение доминантного собственного значения/вектора):
  Доминантное собственное значение: 5.000000285665802
  Соответствующий собственный вектор: [0.70710698 0.70710658]

  Проверка (A * v vs lambda * v):
    A * v: [3.53553451 3.5355337 ]
    lambda * v: [3.53553512 3.5355331 ]
    Разница (норма): 8.569982249548373e-07
------------------------------
Результат с использованием np.linalg.eig (все собственные значения/вектора):
  Все собственные значения: [5. 2.]
  Соответствующие собственные векторы (по столбцам):
[[ 0.70710678 -0.4472136 ]
 [ 0.70710678  0.89442719]]

  Проверка для всех пар (A * v[:, i] vs lambda[i] * v[:, i]):
    Пара 1:
      Собственное значение: 5.0
      Собственный вектор: [0.70710678 0.70710678]
      A * v: [3.53553391 3.53553391]
      lambda * v: [3.53553391 3.53553391]
      Разница (норма): 0.0
----------
    Пара 2:
      Собственное значение: 2.0
      Собств

Модель 2: ChatGPT (OpenAI)

In [22]:
import numpy as np

def power_iteration_chatgpt(A, num_iters=1000, tol=1e-9):
    """
    Степенной метод для поиска доминирующего собственного значения и вектора.
    """
    n, _ = A.shape
    b_k = np.random.rand(n)
    b_k /= np.linalg.norm(b_k)  # Нормализация

    for _ in range(num_iters):
        # Умножаем A на текущий вектор
        b_k1 = A @ b_k

        # Нормализуем
        b_k1_norm = np.linalg.norm(b_k1)
        b_k1 /= b_k1_norm

        # Проверка сходимости
        if np.linalg.norm(b_k1 - b_k) < tol:
            break

        b_k = b_k1

    # Оценка собственного значения
    eigenvalue = (b_k.T @ A @ b_k)  # Релеевское частное
    eigenvector = b_k
    return eigenvalue, eigenvector

def find_all_eigenvalues_vectors_numpy(A):
    """
    Вычисление всех собственных значений и векторов.
    """
    eigenvalues, eigenvectors = np.linalg.eig(A)
    return eigenvalues, eigenvectors

# === Пример использования ===
A = np.array([
    [4, 1],
    [2, 3]
], dtype=float)

# Степенной метод
dom_val, dom_vec = power_iteration(A)
print("Доминирующее собственное значение (степенной метод):", dom_val)
print("Соответствующий собственный вектор:", dom_vec)

# Полная декомпозиция
eigvals, eigvecs = find_all_eigenvalues_vectors_numpy(A)
print("\nВсе собственные значения (numpy):", eigvals)
print("Все собственные векторы (numpy):\n", eigvecs)

Доминирующее собственное значение (степенной метод): 5.0
Соответствующий собственный вектор: [0.70710678 0.70710678]

Все собственные значения (numpy): [5. 2.]
Все собственные векторы (numpy):
 [[ 0.70710678 -0.4472136 ]
 [ 0.70710678  0.89442719]]


Модель 3: DeepSeek

In [23]:
import numpy as np

def power_method_deepseek(A, num_iterations=100, tol=1e-10):
    """
    Вычисляет доминантное собственное значение и соответствующий собственный вектор матрицы A.

    Параметры:
        A : numpy.ndarray
            Квадратная матрица.
        num_iterations : int
            Максимальное число итераций.
        tol : float
            Критерий остановки (изменение нормы вектора).

    Возвращает:
        eigenvalue : float
            Доминантное собственное значение.
        eigenvector : numpy.ndarray
            Соответствующий собственный вектор (нормированный).
    """
    n = A.shape[0]
    eigenvector = np.random.rand(n)  # Начальное приближение
    eigenvector = eigenvector / np.linalg.norm(eigenvector)

    for _ in range(num_iterations):
        old_vector = eigenvector.copy()
        eigenvector = A @ eigenvector
        eigenvector = eigenvector / np.linalg.norm(eigenvector)

        # Проверка сходимости
        if np.linalg.norm(eigenvector - old_vector) < tol:
            break

    eigenvalue = np.dot(eigenvector, A @ eigenvector)
    return eigenvalue, eigenvector

def get_eigen_numpy_deepseek(A):
    """
    Вычисляет все собственные значения и векторы матрицы A.

    Параметры:
        A : numpy.ndarray
            Квадратная матрица.

    Возвращает:
        eigenvalues : numpy.ndarray
            Массив собственных значений.
        eigenvectors : numpy.ndarray
            Матрица, где каждый столбец — собственный вектор.
    """
    eigenvalues, eigenvectors = np.linalg.eig(A)
    return eigenvalues, eigenvectors

# Тестовая матрица 3x3
A = np.array([[4, 1, -1],
              [1, 3, 1],
              [-1, 1, 2]])

# 1. Степенной метод
dom_eigenvalue, dom_eigenvector = power_method_deepseek(A)
print("Доминантное собственное значение (степенной метод):", dom_eigenvalue)
print("Соответствующий собственный вектор:", dom_eigenvector)

# 2. Полное разложение
eigenvalues, eigenvectors = full_eigen_decomposition(A)
print("\nВсе собственные значения (NumPy):", eigenvalues)
print("Все собственные векторы (по столбцам):\n", eigenvectors)

# Проверка: A * v = λ * v для первого собственного вектора
first_eigenvector = eigenvectors[:, 0]
first_eigenvalue = eigenvalues[0]
print("\nПроверка A*v == λ*v:", np.allclose(A @ first_eigenvector, first_eigenvalue * first_eigenvector))

Доминантное собственное значение (степенной метод): 4.6751308705666474
Соответствующий собственный вектор: [ 0.88765034  0.42713229 -0.17214786]

Все собственные значения (NumPy): [0.78568026 4.67513087 3.53918887]
Все собственные векторы (по столбцам):
 [[-0.39711255  0.88765034 -0.23319198]
 [ 0.52065737  0.42713229  0.73923874]
 [-0.75578934 -0.17214786  0.63178128]]

Проверка A*v == λ*v: True


### 6.3. Запуск сниппетов на любой выбранной матрице и численное сравнение с вашей реализацией

In [24]:
A = np.random.randn(4, 4)

print("\n--- Сравнение реализаций на матрице A ---")
print(f"Матрица A:\n{A}")

# Расчет собственных значений и векторов
eigenvalues, eigenvectors = np.linalg.eig(W)

# 1. Собственная реализация (из Раздела 5)
print("\n1. Собственная реализация (Раздел 5):")
print(f"   Все с.з. (numpy.linalg.eig): {np.round(eigenvalues, 6)}")

# 2. Gemini
print("\n2. Реализация Gemini:")
try:
    dom_eigval_g_A, _ = power_iteration_gemini(A)
    all_eigvals_g_A, _ = get_all_eigen_np(A)
    all_eigvals_g_A_sorted = np.sort(np.abs(all_eigvals_g_A))[::-1] # Сортировка по модулю для сравнения
    print(f"   Доминантное с.з. (степенной метод): {dom_eigval_g_A:.6f}")
    print(f"   Все с.з. (numpy.linalg.eig): {np.round(np.sort(all_eigvals_g_A)[::-1], 6)}") # Сортируем для наглядности
    print(f"   Разница доминантных с.з.: {np.abs(dominant_eigenvalue - dom_eigval_g_A):.2e}")
    print(f"   Разница всех с.з. (норма): {np.linalg.norm(np.sort(eigenvalues_np) - np.sort(all_eigvals_g_A)):.2e}")
except Exception as e:
    print(f"   Ошибка при выполнении кода Gemini: {e}")


# 3. ChatGPT
print("\n3. Реализация ChatGPT:")
try:
    dom_eigval_c_A, _ = power_iteration_chatgpt(A)
    all_eigvals_c_A, _ = find_all_eigenvalues_vectors_numpy(A)
    all_eigvals_c_A_sorted = np.sort(np.abs(all_eigvals_c_A))[::-1] # Сортировка по модулю для сравнения
    print(f"   Доминантное с.з. (степенной метод): {dom_eigval_c_A:.6f}")
    print(f"   Все с.з. (numpy.linalg.eig): {np.round(np.sort(all_eigvals_c_A)[::-1], 6)}")
    print(f"   Разница доминантных с.з.: {np.abs(dominant_eigenvalue - dom_eigval_c_A):.2e}")
    print(f"   Разница всех с.з. (норма): {np.linalg.norm(np.sort(eigenvalues_np) - np.sort(all_eigvals_c_A)):.2e}")
except Exception as e:
    print(f"   Ошибка при выполнении кода ChatGPT: {e}")


# 4. DeepSeek
print("\n4. Реализация DeepSeek:")
try:
    dom_eigval_gc_A, _ = power_method_deepseek(A)
    all_eigvals_gc_A, _ = get_eigen_numpy_deepseek(A)
    all_eigvals_gc_A_sorted = np.sort(np.abs(all_eigvals_gc_A))[::-1] # Сортировка по модулю для сравнения
    print(f"   Доминантное с.з. (степенной метод): {dom_eigval_gc_A:.6f}")
    print(f"   Все с.з. (numpy.linalg.eig): {np.round(np.sort(all_eigvals_gc_A)[::-1], 6)}")
    print(f"   Разница доминантных с.з.: {np.abs(dominant_eigenvalue - dom_eigval_gc_A):.2e}")
    print(f"   Разница всех с.з. (норма): {np.linalg.norm(np.sort(eigenvalues_np) - np.sort(all_eigvals_gc_A)):.2e}")
except Exception as e:
    print(f"   Ошибка при выполнении кода DeepSeek: {e}")


--- Сравнение реализаций на матрице A ---
Матрица A:
[[-0.54438272  0.11092259 -1.15099358  0.37569802]
 [-0.60063869 -0.29169375 -0.60170661  1.85227818]
 [-0.01349722 -1.05771093  0.82254491 -1.22084365]
 [ 0.2088636  -1.95967012 -1.32818605  0.19686124]]

1. Собственная реализация (Раздел 5):
   Все с.з. (numpy.linalg.eig): [ 1.437505+0.j       -0.79408 +0.j       -0.703277+1.066861j
 -0.703277-1.066861j]

2. Реализация Gemini:
   Доминантное с.з. (степенной метод): 1.751437
   Все с.з. (numpy.linalg.eig): [ 1.751437+0.j      -0.380741+1.67338j -0.380741-1.67338j
 -0.806625+0.j     ]
   Разница доминантных с.з.: 2.92e+00
   Разница всех с.з. (норма): 1.02e+00

3. Реализация ChatGPT:
   Доминантное с.з. (степенной метод): 1.751437
   Все с.з. (numpy.linalg.eig): [ 1.751437+0.j      -0.380741+1.67338j -0.380741-1.67338j
 -0.806625+0.j     ]
   Разница доминантных с.з.: 2.92e+00
   Разница всех с.з. (норма): 1.02e+00

4. Реализация DeepSeek:
   Доминантное с.з. (степенной метод): 1.66

### 6.4. Анализ полученных результатов

| Модель        | Комментарий                                                                                                                                                                                                                                                                                       |
| ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Gemini (Google) | Предоставил рабочие функции как для степенного метода, так и для использования `np.linalg.eig`. Код хорошо структурирован, включает комментарии и проверку результатов (A\*v ≈ λ\*v), что очень полезно. Реализация степенного метода включает критерий остановки по толерантности.          |
| ChatGPT (OpenAI) | Также предоставил корректные и рабочие функции. Код лаконичен и соответствует запросу. Степенной метод реализован с критерием остановки.                                                                                                                                               |
| DeepSeek      | Сгенерировал рабочий код для обоих методов. Включил проверку `np.allclose` для `np.linalg.eig`, что является хорошей практикой. Использовал другую тестовую матрицу, но функции универсальны. Изначально в коде была ошибка вызова функции `full_eigen_decomposition(A)`, которая не была определена в сниппете, пришлось заменить на `get_eigen_numpy_deepseek(A)`. |

Все три модели успешно сгенерировали код для нахождения собственных значений и векторов. Небольшие отличия заключаются в стиле кода, комментариях и способах проверки. Реализации степенного метода немного различаются (например, в способе оценки собственного значения и проверке сходимости), но все они следуют основной идее алгоритма. Код от DeepSeek потребовал минимального исправления из-за ошибки в вызове функции.


---

## Раздел 7. Оптимизация LLM при помощи метода главных компонент

### Справка: Метод главных компонент простыми словами с акцентом на собственные числа и собственные векторы

**Определение**

Метод главных компонентов (PCA - Principal Component Analysis) это подход к уменьшению размерности, такой что:
1. Находит новые оси (основные компоненты), где данные отличаются больше всего.
2. Проецирует данные на меньшее количество измерений, сохраняя при этом максимальную дисперсию.

**Как это работает с позиции собсвенных чисел и векторов:**
1. Берем матрицу данных `X` (например, включения токенов или скрытые слои LLM).
2. Вычисляем ковариационную матрицу:  
   $$
   C = \frac{1}{n} X^T X
   $$
3. Находим собственные числа и собственные векторы ковариационной матрицы`C`.
   - *Собственные векторы* = направления (главные компоненты)
   - *Собственные числа* = величина отклонения (дисперсии), зафиксированная в каждом направлении
4. Выберем только *топ-k собственных векторов* (наибольшие собственные числа) → получим сокращённое число измерений.

**Значение метода главных компонент для оптимизации размера LLM**

1. Сжатия слоев встраивания (embedding kayers) или скрытых представлений
- Пример: Вы сокращаете 768-мерные встраивания BERT до 128D, используя 128 лучших собственных векторов → огромная экономия памяти.
- Полезно для **дистилляции**, **квантования** или **работе на устройстве**.

2. Удалить избыточность в пространствах внимания (attention matrix)/фичах (Feature spaces)
- Заголовки внимания часто имеют "низкоуровневую структуру" (например, всего несколько доминирующих собственных мод).
- PCA может выявлять и "сокращать" или объединять заголовки, которые обладают сходными характеристиками.

3. Проанализировать репрезентативную способность
Построение собственного значения **спектра** скрытых состояний говорит вам:
- Насколько выразительная часть модели сосредоточена всего в нескольких направлениях.
- Является ли результат работы слоя **низкоуровневым** (сжимаемым).


### Визуальное представление
- **Плоская кривая собственных значений** → разнообразные, высокоуровневые представления (сложнее сжимать).
- **Крутой спад собственных значений** → большая часть информации сосредоточена в нескольких направлениях (легко сжимается с помощью PCA).

### Реальное приложение:
Google, Meta и HuggingFace использовали PCA и **низкоуровневую аппроксимацию** (например, усеченный SVD, линейные узкие места) для:
  - Сжатия моделей для мобильного развертывания.
  - Ускорения вывода.
  - Понять, на что модель "обращает внимание".



**PCA использует собственные значения/векторы ковариационной матрицы для определения доминирующих направлений в данных. В LLMs PCA выявляет избыточную структуру и обеспечивает сжатие путем проецирования вложений/выходных данных в подпространство меньшей размерности, что делает модели более быстрыми и компактными без существенной потери точности.**

7.1. Протокол запросов к выбранному (одному) ассистенту. Вы должны убедить его предоставить вам код оптимизации модели LLM (с которой Вы начинали работать в разделах 3-4), использующий метод главных компонентов с акцентом на использование методов расчёта собственных значений и собсвенных векторов. Также необходимо протестировать точность до и после оптимизации при помощи сокращённого набора данных (например, Tiny BERTScore (50 примеров)). Для оптимизации слой можно использовать любой, какой выберете или какой насоветует AI. Не обязательно сокращать все слои.

7.2. Полученный в результате код (обновляете в процессе работы над этим разделом, можно оставить только то, что получится в итоге)

7.3. Анализ трудностей и галлюцинаций на этом этапе. Как вы с этим справлялись (выявляли и исправляли).

7.4. Анализ полученной оптимизированной модели и применения метода

### 7.1. Протокол запросов к выбранному ассистенту.

Укажите, какой выбрали в итоге AI-ассистент. Приведите список промптов, которые вы пробовали.

Выбранный ассистент: **Gemini (Google)**

Промпт 1:

Я хочу уменьшить размерность матрицы весов одного из слоев LLM с помощью метода главных компонент (PCA). Мне нужно сохранить, скажем, 95% дисперсии. Покажи, как это сделать на Python, используя NumPy или Scikit-learn. Нужно:
1. Взять матрицу весов (предположим, она в переменной 'target_matrix').
2. Выполнить PCA (или использовать SVD, что тесно связано с PCA для ковариационной матрицы).
3. Определить количество компонент, необходимое для сохранения 95% дисперсии.
4. Получить матрицу проекции (компоненты PCA).
5. Опционально: показать, как можно было бы спроецировать исходную матрицу на пространство меньшей размерности и восстановить ее (с потерей информации).

### 7.2. Полученный код (в результате совместной работы и доработок):

In [33]:
import numpy as np
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt # Для опциональной визуализации

# --- Шаг 1: Загрузка или создание матрицы весов ---
# Предположим, у нас есть матрица весов 'target_matrix'.

print(f"Исходная размерность матрицы: {target_matrix.shape}")

# --- Шаг 2: Выполнение PCA ---
# Мы хотим сохранить 95% дисперсии.
# Scikit-learn PCA позволяет указать долю дисперсии как n_components (от 0.0 до 1.0).
# PCA в scikit-learn по умолчанию центрирует данные (вычитает среднее).
# Это важно для корректной работы PCA.
variance_to_keep = 0.95
pca = PCA(n_components=variance_to_keep)

# Обучаем PCA на нашей матрице.
# В контексте PCA, каждая строка матрицы рассматривается как отдельный "образец",
# а столбцы - как "признаки".
pca.fit(target_matrix)

# --- Шаг 3: Определение количества компонент ---
# После обучения pca.n_components_ будет содержать количество компонент,
# необходимое для сохранения >= 95% дисперсии.
n_components_selected = pca.n_components_
explained_variance_ratio = pca.explained_variance_ratio_
total_explained_variance = np.sum(explained_variance_ratio)

print(f"\n--- Результаты PCA ---")
print(f"Необходимое количество компонент для сохранения {variance_to_keep*100:.1f}% дисперсии: {n_components_selected}")
print(f"Фактически сохраненная доля дисперсии: {total_explained_variance:.4f} ({total_explained_variance*100:.2f}%)")
print(f"Доля дисперсии, объясняемая каждой компонентой: {explained_variance_ratio}")

# Можно посмотреть на кумулятивную сумму дисперсии
# cumulative_variance_ratio = np.cumsum(explained_variance_ratio)
# plt.figure(figsize=(8, 5))
# plt.plot(range(1, n_components_selected + 1), cumulative_variance_ratio, marker='.')
# plt.xlabel('Количество компонент')
# plt.ylabel('Кумулятивная доля объясненной дисперсии')
# plt.title('Зависимость объясненной дисперсии от количества компонент')
# plt.grid(True)
# plt.axhline(y=variance_to_keep, color='r', linestyle='--', label=f'{variance_to_keep*100:.0f}% порог')
# plt.legend()
# plt.show()


# --- Шаг 4: Получение матрицы проекции ---
# Главные компоненты (оси нового пространства) хранятся в pca.components_.
# Это и есть матрица проекции. Она имеет размерность (n_components, n_features),
# где n_features - исходное количество столбцов (768 в нашем примере).
# Каждая строка этой матрицы - это одна главная компонента (собственный вектор ковариационной матрицы).
projection_matrix = pca.components_
print(f"\nРазмерность матрицы проекции (главные компоненты): {projection_matrix.shape}")
# Эта матрица используется для преобразования данных из исходного пространства
# (768 измерений) в пространство меньшей размерности (n_components_selected измерений).
# Проекция: Z = X @ projection_matrix.T  (где X - центрированные данные)
# В Scikit-learn это делает метод transform().


# --- Шаг 5 (Опционально): Проекция и восстановление ---

# 5а. Проецирование исходной матрицы на пространство меньшей размерности
# Метод transform() центрирует данные и умножает их на projection_matrix.T
projected_matrix = pca.transform(target_matrix)
print(f"\n--- Проекция и Восстановление ---")
print(f"Размерность спроецированной матрицы: {projected_matrix.shape}") # Будет (128, n_components_selected)

# 5б. Восстановление матрицы (с потерей информации)
# Метод inverse_transform() выполняет обратное преобразование:
# Восстановленное = Спроецированное @ projection_matrix + Среднее
reconstructed_matrix = pca.inverse_transform(projected_matrix)
print(f"Размерность восстановленной матрицы: {reconstructed_matrix.shape}") # Вернется к (128, 768)

# Оценка ошибки восстановления
reconstruction_error = np.linalg.norm(target_matrix - reconstructed_matrix, 'fro')
original_norm = np.linalg.norm(target_matrix, 'fro')
relative_error = reconstruction_error / original_norm
print(f"Норма Фробениуса исходной матрицы: {original_norm:.4f}")
print(f"Норма Фробениуса ошибки восстановления: {reconstruction_error:.4f}")
print(f"Относительная ошибка восстановления: {relative_error:.4f} ({relative_error*100:.2f}%)")

# Важное замечание о связи с SVD:
# PCA тесно связан с сингулярным разложением (SVD). Если вы выполните SVD
# центрированной матрицы X_centered = U * S * Vt, то строки матрицы Vt
# (т.е. V.T) будут главными компонентами (собственными векторами X_centered.T @ X_centered),
# а квадраты сингулярных чисел S, поделенные на (количество_образцов - 1),
# дадут объясненную дисперсию (собственные значения ковариационной матрицы).
# Scikit-learn PCA часто использует SVD "под капотом" для вычислений, особенно для больших матриц.

Исходная размерность матрицы: (768, 768)

--- Результаты PCA ---
Необходимое количество компонент для сохранения 95.0% дисперсии: 388
Фактически сохраненная доля дисперсии: 0.9503 (95.03%)
Доля дисперсии, объясняемая каждой компонентой: [0.01286784 0.01128501 0.01069112 0.00932693 0.00797869 0.00786707
 0.00778011 0.00745845 0.00741655 0.00734801 0.00727699 0.00716053
 0.0071362  0.00698394 0.00694198 0.00682403 0.00673526 0.00672915
 0.00662187 0.00655235 0.00650778 0.00646553 0.00643839 0.00631162
 0.00628755 0.00622782 0.00619011 0.0061369  0.00606686 0.00597847
 0.00590639 0.00581772 0.0057987  0.00574746 0.00567227 0.00564209
 0.00559153 0.00553519 0.00552118 0.00538841 0.00535818 0.0052699
 0.00523871 0.00517568 0.00511436 0.00509587 0.00500868 0.00495467
 0.00493427 0.00485945 0.00483432 0.00475583 0.00468894 0.00462882
 0.00460698 0.00454504 0.00453372 0.00448768 0.0044521  0.00443664
 0.00437881 0.00429182 0.00424983 0.00419139 0.00412667 0.00408764
 0.00404979 0.00404748 0.00

### 7.3. Анализ трудностей и галлюцинаций на этом этапе. Как вы с этим справлялись (выявляли и исправляли).

При работе с Gemini над задачей применения PCA к матрице весов LLM основные трудности были связаны не столько с галлюцинациями в коде (код для PCA с использованием `sklearn` был в целом корректен), сколько с необходимостью точно сформулировать запрос и интерпретировать результат в контексте LLM.

В целом, галлюцинации на этом этапе были минимальны, но потребовалось внимательное чтение кода и осмысление результатов PCA в специфическом контексте весовых матриц LLM.

### 7.4. Анализ полученной оптимизированной модели и применения метода

Применение метода главных компонент (PCA) к матрице весов `q_weights` размером `(768, 768)` для сохранения 95% дисперсии позволило сократить размерность до 388 компонент. Это показывает, что значительная часть вариативности исходного 768-мерного пространства может быть представлена в меньшем, 388-мерном подпространстве. Тот факт, что для 95% дисперсии потребовалась примерно половина исходных измерений, указывает на наличие избыточности в данных, при этом информация распределена по многим компонентам, а не сконцентрирована в нескольких доминирующих. Однако такое сжатие сопряжено с относительной ошибкой восстановления около 22.27%, что означает заметную потерю информации при реконструкции матрицы из главных компонент. Хотя результаты PCA, такие как матрица проекции и спроецированные данные, потенциально могут быть использованы для замены исходного линейного слоя в LLM на два меньших слоя для сокращения параметров, необходимо тщательно валидировать производительность измененной модели из-за существенной ошибки восстановления, которая может негативно повлиять на качество предсказаний. В целом, PCA выявил возможность сжатия матрицы весов, но его практическое применение требует оценки влияния на конечную производительность модели.

---

Что хочется добавить к сегодняшнему занятию:

1. [Код](https://github.com/tongjingqi/MathTrap) со ссылкой на статью про математические ловушки для LLM

2. [Заметка](https://habr.com/ru/articles/904754/) о том, какие навыки *кажется следует* подтягивать программистам в современной индустрии

3. Работая с различными AI-ассистентами, по крайней мере в Университете, будьте готовы к тому, что Вы не сможете получить максимум баллов за выполненные с их помощью работы, если только в задании не требуется обратное. Я провела опрос среди преподавателей нашего факультета (участие приняли 30 человек). 50% преподавателей высказались о том, что не поставят максимальный балл за работу выполненную с помощью GPT и аналогов, предлагая урезать баллы минимум в половину. Ещё 30% процентов поставят 0 баллов за такую работу и только лишь 20% преподавателей согласны принять работу выполненную при помощи ИИ-агентов и оценивать как выполненную самостоятельно.

## Итог

Расскажите о Вашем сегодняшнем опыте: если это привычный способ изучения новой области, прототипирования, то на сколько удалось погрузиться в тему собственных чисел и оптимизации LLM. Если вообще все Вам показалось новым, включая общение с LLM, то расскажите, как Вы справлись с написанием промптов и кода? Если писали пропты на языке отличном от родного, то на сколько трудно это Вам показалось и какие подходы использовали (например, машинный перевод?)

Как Вы сами оцениваете свой прогресс на сегодняшнем занятии?

Удалось ли Вам изучить всё то, что перечисленно в самом первом списке в ноутбуке? Что вызвало наибольшее затруднение?

---

Опыт изучения собственных чисел, векторов и PCA для оптимизации моделей с помощью LLM (Gemini, ChatGPT, DeepSeek) был успешным.

AI-ассистенты помогли быстро освоить основные концепции и получить готовый код на Python. Ключом было написание точных запросов с указанием всех деталей (библиотеки, методы).

Сгенерированный код в основном работал, но иногда требовал небольших исправлений. Ошибки находились при запуске и проверке.

Самой большой сложностью было не написание кода, а понимание результатов (например, почему PCA давал такие результаты на весах моделей) и исправление мелких ошибок AI.

В целом, AI значительно ускорил процесс обучения и экспериментов, но требует критической проверки результатов и понимания предметной области. Это хороший инструмент для быстрого старта, но не замена глубоким знаниям.